### DISCUSSION TITLE
A way to trigger Web Socket `onerror` before `connection` event

### DISCUSSION THREAD

**Discussion by sfc-gh-jtjhin** [mswjs/msw#2364] [Q&A]
### Scope

Improves an existing behavior

### Compatibility

- [ ] This is a breaking change

### Feature description

I'd like to be able to test if the a web socket connection erroring e.g. firewall issue
Similar to https://mswjs.io/docs/recipes/network-errors for REST requests

**Comment by kettanaito** [comment]
Hi, @sfc-gh-jtjhin. Yes, you can emulate the `error` event on a WebSocket. That event represents connection errors, and to achieve it you can throw an error in the `connection` event listener:

```ts
const api = ws.link('wss://localhost')

export const handlers = [
  api.addEventListener('connection', () => {
    throw new Error('Connection error')
  })
]
```

This will get translated to a connection error, and will trigger both `close` and `error` events per spec ([implementation reference](https://github.com/mswjs/interceptors/blob/0ce43e8f6b97f1fe6880c6ea2789fb50ffc9e78d/src/interceptors/WebSocket/index.ts#L129-L150)).

This works because the `connection` event is dispatched while your WebSocket client is still in the `CONNECTING` ready state. That is precisely for you to be able to error that connection before it is open. If you don't error it, the (mocked) connection will open at the next event loop tick.

**Reply by sfc-gh-jtjhin** [reply]
Okay weirdly it is not hitting my `ws.onerror = () => {}` callback but `onopen` instead.

```
    const handler = this.link.addEventListener('connection', ({ client, info }) => {
      throw new Error('Connection error');
    });
    server.use(handler)
```

But I created a simplified test and it does hit `onerror` 🤷‍♂️

**Reply by sfc-gh-jtjhin** [reply]
@kettanaito  Actually I can't get it to fail on a simplified test:

```
import { ws } from 'msw';
import { setupServer } from 'msw/node';

describe('Web socket test', () => {
  beforeEach(() => {
    const api = ws.link('wss://example.com/path');

    const server = setupServer(
      api.addEventListener('connection', () => {
        console.log('connecting');
        throw new Error('Connection error');
      }),
    );
    server.listen();
  });

  it('Connection should error', (done) => {
    const connection = new WebSocket('wss://example.com/path');
    connection.onerror = (e) => {
      console.log('onerror', e);
      done();
    };

    connection.onclose = (e) => {
      console.log('onclose', e);
    };

    connection.onopen = () => {
      console.log('onopen');
      throw new Error('Connection should not be opened');
    };
  });
});
```

this logs:

```
connecting...
onopen
```

But never reaches `onerror`

**Reply by sfc-gh-jtjhin** [reply]
Okay I put a bunch of console logs in and the error thrown from the connection handler is definitely getting eaten by the event emitter and not hitting the `catch` block.

**Reply by kettanaito** [reply]
@sfc-gh-jtjhin, can you please put it in a standalone repo for me to reproduce? That'd be great.

**Reply by joshua-tj** [reply]
@kettanaito here you go https://github.com/joshua-tj/msw-ws-onerror-test/

**Reply by sfc-gh-jtjhin** [reply]
@kettanaito were you able to reproduce it with the repo?

**Comment by sfc-gh-jtjhin** [comment]
Re-created bug report https://github.com/mswjs/msw/issues/2446
